### PR TITLE
Add water mask material

### PR DESCRIPTION
**Pull request by ggetz**
# Description

Adds a new water mask globe material. 

image

The motivation is so that custom globe materials can be written that shade only the land or only the sea, as demonstrated in the [new Sandcastle example](https://ci-builds.cesium.com/cesium/water-mask-globe-material/Apps/Sandcastle/index.html?src=Globe%20Materials%20%E2%80%93%20Water%20Mask%20Elevation%20Map.html).

image

In order to do this, I separated the concerns of having a water mask from displaying the water effect in `packages/engine/Source/Scene/GlobeSurfaceShaderSet.js`, and added the new material shader.

## Issue number and link

N/A

## Testing plan

1. Ensure the [new Sandcastle example](https://ci-builds.cesium.com/cesium/water-mask-globe-material/Apps/Sandcastle/index.html?src=Globe%20Materials%20%E2%80%93%20Water%20Mask%20Elevation%20Map.html) works. 
2. Alter the properties of the `customElevationMaterial` (such as by using the `waterMaskMaterial.diffuse` value) to see how the mask interacts with the diffuse channel.
3. Test the uniforms by setting

```js
material.materials.waterMaskMaterial.uniforms.landColor = Cesium.Color.YELLOW.clone();
material.materials.waterMaskMaterial.uniforms.waterColor = Cesium.Color.BLUE.clone();
```

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

### CODE REVIEW

**Comment by ggetz** [comment]
@jjspace A reminder to review this PR.

**Review by jjspace — APPROVED** [review]
Sorry for the delay getting back to this. As far as I can tell (with my more limited knowledge of shaders/materials) all the code looks good. I was also able to test and the sandcastles work as expected.

I think the only thing I might add is that it could be nice to demonstrate the testing plan you outlined in the new sandcastle example. Just a simple material picker like [this](https://ci-builds.cesium.com/cesium/water-mask-globe-material/Apps/Sandcastle/index.html#c=1Vdbb9s2FP4rhPcwG7ApO3GWJk6KoZmxPrgt0KQNhroPtHRkEaVIjaTspEP/+w5JXWN3Kfo2GJAt8twv3zmOlTSW7DjsQZNrImFPbsDwMqcf/dlwPYj9+42SlnEJej0Yk3/WkhALWuPJZc1wF95pqlV+r7RIqoOhpyZEw98lGPsRtIWHt0rnTJhLYnUJYxJF5C1AAglJlSYZF8JkLOFy22O9Z6jzDTNfDtmsIgk3FjlKbjIimERRaAjZOx4n5dto7K3mOQj045KkqB/8GZM8Z5Yr2R5+Gy3Wci1RwY0GFEEY2Qq1AZI7eZwJb2hlJAEBOy+AKCke8eENWMvYRzcujVX5sqZ5U0vohbs+DdFK2Ubz+JJUobOPBRq8HjQyVk78YBxua5NMQ0+C2y5UtdzOXSuvCWgjC+PU/Gq8es/y4j/lLLuUR2TV37HKCyVB2q6pCU/T0ng5RzXSiqAjl4kiY45hRqdkcugs9QSuUjGB9/CrEBhqVyUZEC53oA0QlfpXz5sjryshl0+octfaHZ5WM2lEGaP1B2US8mz36g69+5/k90dzchjcJh+VoB+J0ZPY0MYneii/lBybKzfUJeJGCeWQqQqjf6V/LVerd/c0Fihy+NPi/dVR+a9WH5Yd6XWCcy5fA99mFukn89mcTheuwFhRaPXgIARIAiwhBljbPA0ze2iYX5yfnx9hBqxMhLlD3l5fIP+nKZ2OCT7mZ+4rPMMDn9gUn9HqtJSxx6QtWO+WYx6OQpKDXB3EJSouc8wZjT3WYRW6Nwf9TO6YWQ9cEIinpnue2Ax5ZtNpe5jVjs0WrfDYPuCJv/cWSIR9J/QkCQJbyh0TCO9I3POzR7LVCd6jyMrGFUI4039qxF9nqovG2NmE37VsDO4tAMmsLcxlFG25ocay+As8xBmTW6BY+ZGfK6jQRCdn04uLKM6UMgjpkxgjVuqJM39i1QTrfYJVM+mkhjijKEtCid5aVQyDI5+mn8fYOL+czufp6ayO3veIZ4H4bPPifH7yHPFJIIbfEsbOniM+DcRpEp+fP2vGPBBvLi6SFxfPEZ9Vkqfu08kmpifF2X1rHwVgupB90T1/D3EvVbOaUYMttfS1giff2o4LM/e6WlCoQWQB6g8XbV81cPudQduQVuP6QwUA5Nrpzg8x4/gYqnEDxT0RRBEYeF7mTX83QHGMlj30aWtcOEKLsLD1gex1sA+aDwLtOJ+33oY7k6m9HwDLNMXAI4lH5OYeJNsIbCRU7XaYa79TLVrJ3sxbbDeQtwWLYam1R0oEGY9cK4VQlyE7bo4IW0qUHm0sF9jLbjvKlXaAiGujwNnbLHR+q/Jq/QD2OSV7jqjCms0PsTAYbXjOBdNuONeNjHhJmY5dP7sOBulOJoijkUBvjY2sUmKiIUUwlTFEp8mESSYe8QYjMmlUTPZKfzE0s3lTH96UfrW5rPiqE1W+OjP9D67BQywT3p8w25P69LJLe8Nw7TWcydPhzHfAdOS3kw9FglYnHvsfcWVlOXS3z3DtAxUMKBCefJw3zC0x0l/FyKVZ64ZmNs7Cit23uGPFqPGs0PAeZIIeY68vd4inK1yl8UYPmxky9KRjvz73RogpdYrF0SiraksIXhjFE4R+hSXA49su3TBsDkF9MJ7Wjt3fjOvbjhvuaNQZLhK22KG7Vu+Bg9STQK2qq3585KxV0Ek3bXJ5XIPn5F9rJYcUGNBhvVP1TW72sV4QtNPaROCIiWjkuBu8J2bWXlQFVFeEKvyMm7nVwZHUaycO5N7a2f/7Uu+MyImIGLuVrimHUbsrHiDR92H46U79xI67vSJujSNhKfOCf9qKJyvhgXa3It2imzEzVoDL1B3ixobpNyDLYR2x0WIwHlwZN9Be1qp+57gxa0tKLYaURhbywiNPtClxu8AtyphRpe4q6rJeJXxHeHJ95D81iQUzBm/SEucnltR68PIqQvoDVqE8jr5DvBDs0ZFls5ercEgpvYrw9TinDf4hS+1Je/dVqfxVaS36/FRz5/e/) for people (like me) that are less experienced with our Material api to know how to do that.

**Comment by ggetz** [comment]
Thanks @jjspace!

> I think the only thing I might add is that it could be nice to demonstrate the testing plan you outlined in the new sandcastle example. Just a simple material picker like [this](https://ci-builds.cesium.com/cesium/water-mask-globe-material/Apps/Sandcastle/index.html#c=1Vdbb9s2FP4rhPcwG7ApO3GWJk6KoZmxPrgt0KQNhroPtHRkEaVIjaTspEP/+w5JXWN3Kfo2GJAt8twv3zmOlTSW7DjsQZNrImFPbsDwMqcf/dlwPYj9+42SlnEJej0Yk3/WkhALWuPJZc1wF95pqlV+r7RIqoOhpyZEw98lGPsRtIWHt0rnTJhLYnUJYxJF5C1AAglJlSYZF8JkLOFy22O9Z6jzDTNfDtmsIgk3FjlKbjIimERRaAjZOx4n5dto7K3mOQj045KkqB/8GZM8Z5Yr2R5+Gy3Wci1RwY0GFEEY2Qq1AZI7eZwJb2hlJAEBOy+AKCke8eENWMvYRzcujVX5sqZ5U0vohbs+DdFK2Ubz+JJUobOPBRq8HjQyVk78YBxua5NMQ0+C2y5UtdzOXSuvCWgjC+PU/Gq8es/y4j/lLLuUR2TV37HKCyVB2q6pCU/T0ng5RzXSiqAjl4kiY45hRqdkcugs9QSuUjGB9/CrEBhqVyUZEC53oA0QlfpXz5sjryshl0+octfaHZ5WM2lEGaP1B2US8mz36g69+5/k90dzchjcJh+VoB+J0ZPY0MYneii/lBybKzfUJeJGCeWQqQqjf6V/LVerd/c0Fihy+NPi/dVR+a9WH5Yd6XWCcy5fA99mFukn89mcTheuwFhRaPXgIARIAiwhBljbPA0ze2iYX5yfnx9hBqxMhLlD3l5fIP+nKZ2OCT7mZ+4rPMMDn9gUn9HqtJSxx6QtWO+WYx6OQpKDXB3EJSouc8wZjT3WYRW6Nwf9TO6YWQ9cEIinpnue2Ax5ZtNpe5jVjs0WrfDYPuCJv/cWSIR9J/QkCQJbyh0TCO9I3POzR7LVCd6jyMrGFUI4039qxF9nqovG2NmE37VsDO4tAMmsLcxlFG25ocay+As8xBmTW6BY+ZGfK6jQRCdn04uLKM6UMgjpkxgjVuqJM39i1QTrfYJVM+mkhjijKEtCid5aVQyDI5+mn8fYOL+czufp6ayO3veIZ4H4bPPifH7yHPFJIIbfEsbOniM+DcRpEp+fP2vGPBBvLi6SFxfPEZ9Vkqfu08kmpifF2X1rHwVgupB90T1/D3EvVbOaUYMttfS1giff2o4LM/e6WlCoQWQB6g8XbV81cPudQduQVuP6QwUA5Nrpzg8x4/gYqnEDxT0RRBEYeF7mTX83QHGMlj30aWtcOEKLsLD1gex1sA+aDwLtOJ+33oY7k6m9HwDLNMXAI4lH5OYeJNsIbCRU7XaYa79TLVrJ3sxbbDeQtwWLYam1R0oEGY9cK4VQlyE7bo4IW0qUHm0sF9jLbjvKlXaAiGujwNnbLHR+q/Jq/QD2OSV7jqjCms0PsTAYbXjOBdNuONeNjHhJmY5dP7sOBulOJoijkUBvjY2sUmKiIUUwlTFEp8mESSYe8QYjMmlUTPZKfzE0s3lTH96UfrW5rPiqE1W+OjP9D67BQywT3p8w25P69LJLe8Nw7TWcydPhzHfAdOS3kw9FglYnHvsfcWVlOXS3z3DtAxUMKBCefJw3zC0x0l/FyKVZ64ZmNs7Cit23uGPFqPGs0PAeZIIeY68vd4inK1yl8UYPmxky9KRjvz73RogpdYrF0SiraksIXhjFE4R+hSXA49su3TBsDkF9MJ7Wjt3fjOvbjhvuaNQZLhK22KG7Vu+Bg9STQK2qq3585KxV0Ek3bXJ5XIPn5F9rJYcUGNBhvVP1TW72sV4QtNPaROCIiWjkuBu8J2bWXlQFVFeEKvyMm7nVwZHUaycO5N7a2f/7Uu+MyImIGLuVrimHUbsrHiDR92H46U79xI67vSJujSNhKfOCf9qKJyvhgXa3It2imzEzVoDL1B3ixobpNyDLYR2x0WIwHlwZN9Be1qp+57gxa0tKLYaURhbywiNPtClxu8AtyphRpe4q6rJeJXxHeHJ95D81iQUzBm/SEucnltR68PIqQvoDVqE8jr5DvBDs0ZFls5ercEgpvYrw9TinDf4hS+1Je/dVqfxVaS36/FRz5/e/) for people (like me) that are less experienced with our Material api to know how to do that.

I would generally agree with this feedback. However, for this particular example, I think that distracts from the intended use case and make things a bit more complicated in the code example, which I'd like to avoid if possible.

File: packages/engine/Source/Scene/GlobeSurfaceShaderSet.js [hunk 1]
```diff
@@ -168,30 +169,31 @@ GlobeSurfaceShaderSet.prototype.getShaderProgram = function (options) {
...
+    (cartographicLimitRectangleFlag << 23) |
+    (imageryCutoutFlag << 24) |
+    (colorCorrect << 25) |
+    (highlightFillTile << 26) |
+    (colorToAlpha << 27) |
+    (hasGeodeticSurfaceNormals << 28) |
+    (hasExaggeration << 29) |
+    (showUndergroundColor << 30) |
+    (translucent << 31) |
+    (applyDayNightAlpha << 32);
```
Comment (bbbbx) [inline packages/engine/Source/Scene/GlobeSurfaceShaderSet.js:196]: `(applyDayNightAlpha << 32)` will return 0 or 1, which conflicts with `sceneMode` and may introduce potential bugs, see #12260

**Pull request merged**